File: leetcode/0498-diagonal-traverse/solution.py
from typing import List


class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        m, n = len(mat), len(mat[0])
        result = []

        for d in range(m + n - 1):
            r_lo = max(0, d - (n - 1))
            r_hi = min(d, m - 1)
            rows = range(r_lo, r_hi + 1)
            if d % 2 == 0:
                rows = reversed(rows)
            for r in rows:
                c = d - r
                result.append(mat[r][c])

        return result

File: leetcode/0498-diagonal-traverse/Solution.java
class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        int m = mat.length, n = mat[0].length;
        int[] result = new int[m * n];
        int idx = 0;

        for (int d = 0; d < m + n - 1; d++) {
            int rLo = Math.max(0, d - (n - 1));
            int rHi = Math.min(d, m - 1);

            if (d % 2 == 0) {
                for (int r = rHi; r >= rLo; r--) {
                    result[idx++] = mat[r][d - r];
                }
            } else {
                for (int r = rLo; r <= rHi; r++) {
                    result[idx++] = mat[r][d - r];
                }
            }
        }

        return result;
    }
}
